### PR TITLE
Tests for tiles + support contentlistingtiles without collections

### PR DESCRIPTION
**Pull request by djay**
This now makes the flow of creating a custom search page using mosaic much easier
- edit mosaic page
- add listingtile (if layout doesn't have by default)
- add filter (should default to the right target css for ajax).
- save.

It still supports using a filter tile targeting a collection but it's not clear how you would get the collection results onto the page when using ajax. You might still be able to use the "existing content" tile?

In addition this adds layer to test tiles. ie every robot test for portlets now becomes a test for tiles also. If you want to test just portlets or tiles use --layer tiles or --layer Portlet.

In addition this switches the settings to using the select2 widget. It's more tidy (and works in 5.0 when used in a tile).

Deco/Mosaic was designed to remove the need for collection objects completely. The contentlisting tile does everything a collection does (except syndication) so there isn't a need for it.

This does perhaps make the name collectionfilter confusing however. #135 aims to fix the names and simplify the settings panels.

Without further changes to contentlisting tile 5.0 is not supported when using tiles (because the change to standardtiles to allow this tile to listen to the request was made non compatible with 5.0 due to dependencies).

### CODE REVIEW

**Comment by djay** [comment]
As part of this I want to simplify the wording and names to try and make things less confusing. esp the order of the fields in popups.

Screen Shot 2021-06-19 at 12 15 15 am

**Comment by djay** [comment]
The default content selector also needs to change. 
Either tiles always default to ```.contentlisting-tile```, or it defaults to blank and it will automatically work out the selector based on what its pointed at.

**Comment by djay** [comment]
@petschki The original reason for the PR was to have tiles tested so we can get other PRs merged to reduce our technical debt. If the inclusion of the contentlisting stuff is going to cause a delay I can potentially split the PR and change the tests to use collections.

File: src/collective/collectionfilter/filteritems.py [hunk 1]
```diff
@@ -9,7 +9,12 @@
 from collective.collectionfilter.vocabularies import DEFAULT_FILTER_TYPE
 from collective.collectionfilter.vocabularies import EMPTY_MARKER
 from Missing import Missing
+from plone.app.blocks.layoutbehavior import ILayoutBehaviorAdaptable
```
Comment (petschki) [inline src/collective/collectionfilter/filteritems.py:12]: I just tried your branch: if you do not have installed plone.app.mosaic (which is no hard dependency) this import breaks ... I'd put it into the try/except block

File: src/collective/collectionfilter/filteritems.py [hunk 2]
```diff
@@ -9,7 +9,12 @@
...
 from plone.app.contenttypes.behaviors.collection import ICollection
+try:
+    from plone.app.blocks.layoutbehavior import ILayoutAware
+except ImportError:
+    ILayoutAware = None
```
Comment (petschki) [inline src/collective/collectionfilter/filteritems.py:17]: Instead of try/except imports I'd refactor the whole `ICollectionish` logic into the tiles folder which is included onyl if `plone.tiles` is present ... collective.collectionfilter doesn't (shouldn't) have a hard dependency on plone.app.mosaic.

**Review by petschki — CHANGES_REQUESTED** [review]
In general I like the approach and the simplification for mosaic filtering but right now, there's a hard dependency to plone.app.blocks which should not be in this package. Please refactor into the `tiles` folder which is only included when `plone.tiles` is installed.

**Comment by petschki** [comment]
> @petschki The original reason for the PR was to have tiles tested so we can get other PRs merged to reduce our technical debt. If the inclusion of the contentlisting stuff is going to cause a delay I can potentially split the PR and change the tests to use collections.

@djay that would be a great idea 👍🏼

**Comment by djay** [comment]
> > @petschki The original reason for the PR was to have tiles tested so we can get other PRs merged to reduce our technical debt. If the inclusion of the contentlisting stuff is going to cause a delay I can potentially split the PR and change the tests to use collections.
> 
> @djay that would be a great idea 👍🏼

I'm saying I'd rather not because its extra work. Is the change I've made that big? if the current tests don't show it works then perhaps thats a bigger problem?

File: src/collective/collectionfilter/filteritems.py (hunk 1 above)
Comment (djay) [reply src/collective/collectionfilter/filteritems.py:12]: Sorry I fixed that but perhaps in a different branch. easy fix

File: src/collective/collectionfilter/filteritems.py (hunk 2 above)
Comment (djay) [reply src/collective/collectionfilter/filteritems.py:17]: it's not a hard dependency

Comment (djay) [reply src/collective/collectionfilter/filteritems.py]: I don't think it makes the code more readable to split it up between tiles folder and filtereditems but I've done it. esp since all the GEO stuff has been kept togeather yet its not put in its own directory.

**Comment by djay** [comment]
@petschki done making changes.

**Comment by petschki** [comment]
I just tried this with a buildout.coredev 5.2 and it works very nice. Since the `tile_class` is only available since `plone.app.standardtiles==2.4.0` I've made a `extra_install` which sets the correct minimum version requirement. I've fixed also some minor typos in README.

But there's one issue I've had when testing it: If you enable `Mosaic layout` on a page and edit it the first time by adding a contentlisting tile and a filter tile, I got the following exception:
```
2021-06-25 08:19:19,961 ERROR   [Zope.SiteErrorLog:252][waitress-3] 1624601959.96020980.47787753061995 http://localhost:8080/Plone/filter-test/@@add-tile/collective.collectionfilter.tiles.filter
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 162, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 371, in publish_module
  Module ZPublisher.WSGIPublisher, line 274, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module plone.z3cform.layout, line 63, in __call__
  Module plone.z3cform.layout, line 47, in update
  Module plone.app.tiles.browser.base, line 104, in update
  Module plone.z3cform.fieldsets.extensible, line 65, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 145, in update
  Module plone.app.z3cform.csrf, line 22, in execute
  Module z3c.form.action, line 98, in execute
  Module z3c.form.button, line 301, in __call__
  Module z3c.form.button, line 159, in __call__
  Module plone.app.tiles.browser.add, line 50, in handleAdd
  Module z3c.form.group, line 98, in extractData
  Module z3c.form.form, line 148, in extractData
  Module z3c.form.field, line 331, in extract
  Module z3c.form.field, line 303, in _extract
  Module collective.collectionfilter.validator, line 32, in validate
  Module zope.component._api, line 96, in queryAdapter
  Module zope.component.hooks, line 135, in adapter_hook
  Module collective.collectionfilter.tiles, line 68, in __init__
TypeError: No contentlisting tile or Collection found
```

This is due to the transient state of the mosaic page ... after you save the page with a contentlisting tile and edit it again you can add filter tiles as expected ... @djay I think this should be solved before we merge this.

**Comment by djay** [comment]
@petschki hmm. I need to have some tests that run just for tiles only, so can test adding to a page instead of a collection.
Or maybe all tilel tests should be on a page.

This happens when you add a filter to a page before a content listing? or is the content listing not saved at all until teh page is saved?
It's possible we might have to live without validating the context is correct (ie has a content listing tiles) in the case its a mosaic page?

**Comment by petschki** [comment]
This is a `Page` with mosaic enabled and edited the first time ... it makes no difference in which order you add filter/contentlisting tiles ... the filtertile always throws the exception, because https://github.com/collective/collective.collectionfilter/pull/133/files#diff-838484ed15774619566b2b294491ed8475f6d5047d9644d5a1297fed39965c9cR78 is always empty when adding a contentlisting tile the first time.

**Comment by djay** [comment]
@petschki in that case I have to change the validation let you pick a collection or any ILayoutAware even if no listing tile.
So then what should happen? filter tiles show an error or just no options?

**Comment by petschki** [comment]
Maybe remove the validator and show a portal message when the page was saved which says `Please add at least one contentlisting tile to your page to show the results of the filters` ... this could be done with an `IObjectModified` event listener. there might be only a javascript solution for validating the mandatory contentlisting tile during editing in the mosaic editor.

**Comment by djay** [comment]
Good idea. I think I can do it with a status message. At least the user will get some feedback.

**Comment by djay** [comment]
@petschki is CF supposed to work if there is a collection anywhere in the acquaisition path? or only if it's in the direct context?
Seems like the the current validator allows it in aqusition but the other code doesn't?

**Comment by petschki** [comment]
Only direct context makes sense to me ...

**Comment by djay** [comment]
@petschki I read the code wrong. it doesn't validate parents.
I've now put in a warning when you edit a tile that appears in the layout editor. Doing it when you save the layout is a bit harder since I'd then have to work out which tiles are filter tiles. I don't think it adds anything to do that.
It does need tests however. Currently there are no tests when you try to set a target other than context, or set a bad target.
There also aren't tile only tests yet to test the warning

File: src/collective/collectionfilter/tiles/__init__.py [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
+import re
+from zope.component import adapter
+from zope.component import getMultiAdapter
+from zope.component import queryAdapter
+from zope.interface import implementer
+from collective.collectionfilter import _
+from collective.collectionfilter.interfaces import ICollectionish
+from collective.collectionfilter.filteritems import CollectionishCollection
+from plone.app.contenttypes.behaviors.collection import ICollection
+from Products.statusmessages.interfaces import IStatusMessage
```
Comment (petschki) [inline src/collective/collectionfilter/tiles/__init__.py:15]: we already have `plone.api`... no need for this import ...

**Review by petschki — CHANGES_REQUESTED** [review]
I've made a quick test on a `Page` context ... see my comments in the code

File: src/collective/collectionfilter/tiles/__init__.py [hunk 4]
```diff
@@ -36,3 +48,111 @@ def filter_id(self):
...
+        self.context = context
+
+        self.selectContent("")  # get first tile
+
+        if self.tile is None:
+            # Could still be a ILayoutAware collection
+            try:
+                self.collection = ICollection(self.context)
+            except TypeError:
+                raise TypeError("No contentlisting tile or Collection found")
```
Comment (petschki) [inline src/collective/collectionfilter/tiles/__init__.py:71]: The `TypeError` is always raised and you cannot add a filtertile on a  `Page` context

Comment (djay) [reply src/collective/collectionfilter/tiles/__init__.py:71]: Fixed. not sure if there is a better way to tell queryadapter it couldn't adapt.

File: src/collective/collectionfilter/tiles/__init__.py (hunk 3 above)
Comment (djay) [reply src/collective/collectionfilter/tiles/__init__.py]: fixed

File: src/collective/collectionfilter/tiles/__init__.py (hunk 4 above)
Comment (petschki) [reply src/collective/collectionfilter/tiles/__init__.py:70]: erm ... there's a missunterstanding ... `zope.component._api.queryAdapter` has a kw parameter `default=None` so you do not get an Exception when using queryAdapter, simply it returns None if no one found ... you should remove this `raise` in line 71 because when you're editing a Page the first time, you do not have a chance to check if there's a contentlisting tile there ...

**Comment by djay** [comment]
@petschki any idea why we are getting 5.2 errors now in the tests with i18n when using tiles to edit a document?

File: src/collective/collectionfilter/tiles/__init__.py (hunk 4 above)
Comment (djay) [reply src/collective/collectionfilter/tiles/__init__.py:70]: Apparently you do get an exception with queryAdapter if one is raised in __init__ of the adapter. None tells you that no adapter was found, not that there was some problem in constructing the adapter and currently I can't see a way to indicate that an ILayoutAware doesn't have contentlisting other than raising an exception.

Comment (petschki) [reply src/collective/collectionfilter/tiles/__init__.py:70]: ok ... but anyway this doesn't work when you initialy build a mosaic page with filter tiles ... you are not able to add filter tiles until you saved your page first with a contentlisting tile ... 

that's why I'd completely remove the check for existing contentlisting tiles while building/customizing a layout. When saving the layout I'd call an `IObjectModified` event which checks for mandatory contentlisting tile if a filtertile is in `ILayoutAware` adapter ... when it's missing -> status message.

Comment (djay) [reply src/collective/collectionfilter/tiles/__init__.py:70]: Thats what the change is supposed to do. I'll work out why its not working.

**Comment by djay** [comment]
@petschki The problem with the tests on 5.2 tiles is this
```
plone.subrequest - ERROR - Error handling subrequest to http://localhost:54321/plone/testdoc/@@plone.app.standardtiles.field?field=IRichTextBehavior-text&X-Tile-Persistent=yes
Traceback (most recent call last):
  File "/home/runner/buildout-cache/eggs/plone.subrequest-1.9.3-py2.7.egg/plone/subrequest/__init__.py", line 163, in subrequest
    bind=1
  File "/home/runner/buildout-cache/eggs/Zope-4.5.5-py2.7.egg/ZPublisher/mapply.py", line 87, in mapply
    return object(*args)
  File "/home/runner/buildout-cache/eggs/plone.app.standardtiles-2.4.0-py2.7.egg/plone/app/standardtiles/field.py", line 122, in __call__
    return self._wrap_widget(widget.render())
  File "/home/runner/buildout-cache/eggs/plone.app.standardtiles-2.4.0-py2.7.egg/plone/app/standardtiles/field.py", line 95, in _wrap_widget
    if render.rstrip().endswith(u"</html>"):
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 7: ordinal not in range(128)
```
It's to do with the richtexttile when you view a page. So my guess is the test setup is creating the richtext content in the incorrect way? I looked at the docs and it seemed ok to me. Or maybe the test server is not setup for utf8?

File: src/collective/collectionfilter/tiles/__init__.py (hunk 4 above)
Comment (djay) [reply src/collective/collectionfilter/tiles/__init__.py]: ok. I've switched it to not throw exception anymore and use different way to determine if there is a listing available.
And added a test for it.

**Comment by djay** [comment]
@petschki the problem with the tests is the one already identified and still not fixed :(
https://github.com/plone/plone.app.standardtiles/issues/111
is 2.4.0 not supposed to be used with python 2.7?

**Comment by djay** [comment]
@petschki fixed the problem with the test setup. RichTextValue wasn't created correctly.
All tests are passing including the new one to check the warning message appears. Ready to test again

**Comment by petschki** [comment]
@djay I was able to test this locally and it works very well. Especially the automagical linking between filter and contentlisting is a huge improvement!

The only thing I found is, that when you save a page with filtertile and without contentlisting, the status message is shown correctly. When you edit and save the second time (also with filter and without contentlisting) the portal message doen't show up anymore ... 

also if you have a page with filter/contentlisting -> edit it and remove the contentlisting and save -> there is no status message/warning about missing contentlisting.

**Comment by djay** [comment]
> The only thing I found is, that when you save a page with filtertile and without contentlisting, the status message is shown correctly. When you edit and save the second time (also with filter and without contentlisting) the portal message doesn't show up anymore ...

Thats when you save the layout or save the filtertile? I only have an handler on the modify event for the filtertile but if you edit that again it should warn you again?

To have a more general warning system for editing anything on the page or just saving the whole layout then I'd need more clever code that finds all the possible filter tiles tests if they are ok. I'll have a think about it

**Comment by djay** [comment]
@petschki it should now warn you whenever you save it. Please review.

**Review by petschki — APPROVED** [review]
LGTM

**Comment by petschki** [comment]
I'll merge this later today, I just want to test it more in a productive project of us.

**Comment by petschki** [comment]
@djay just to be sure: the refactored UI of the tile-add-form isn't in this PR anymore, right? At least my add-forms are the same as before ...

**Comment by djay** [comment]
@petschki correct. pretty sure all of that was in https://github.com/collective/collective.collectionfilter/pull/135.

**Comment by djay** [comment]
@petschki ah there was one UI change. I switched it to using the select2 widget. https://github.com/collective/collective.collectionfilter/pull/133/files#diff-4c74ac2fd432682735f055c860ac2e0dbc61b47acaed3ee94a1553660707f299
I can't remember exactly why but it was something to do with the robot tests and also because it looks a lot nicer IMO.

**Review by agitator — APPROVED** [review]
lgtm

**Comment by djay** [comment]
@agitator I merge or you?
